bluetooth: gatt: add authorization callback API for gatt operations

kapi-no opened the pull request:
Added the GATT authorization callback API that allows the user to define application-specific authorization logic for GATT operations.

Example usage for the Bluetooth Peripheral with the Device Information Service and high privacy requirements:

```
static bool gatt_uuid_16_read_operation_authorize(struct bt_conn *conn,
						  const struct bt_uuid_16 *uuid)
{
	switch (uuid->val) {
	case BT_UUID_GAP_DEVICE_NAME_VAL:
	case BT_UUID_DIS_MANUFACTURER_NAME_VAL:
	case BT_UUID_DIS_MODEL_NUMBER_VAL:
	case BT_UUID_DIS_FIRMWARE_REVISION_VAL:
		/* Validate if the application state allows to expose
		 * identifying information.
		 */
		return identifying_info_read_allow(conn);
	default:
		return true;
	}
}

static bool gatt_read_operation_authorize(struct bt_conn *conn,
					  const struct bt_gatt_attr *attr)
{
	switch (attr->uuid->type) {
	case BT_UUID_TYPE_16:
		return gatt_uuid_16_read_operation_authorize(
			conn, (struct bt_uuid_16 *) attr->uuid);
	default:
		return true;
	}
}

static const struct bt_gatt_authorization_cb gatt_authorization_callbacks = {
	.read_operation_authorize = gatt_read_operation_authorize,
};

(...)

int main(void)
{
	(...)

	err = bt_gatt_authorization_cb_register(&gatt_authorization_callbacks);
	if (err) {
		LOG_ERR("Registering GATT authorization callbacks failed (err %d)", err);
		return err;
	}

	(...)
}
```

kapi-no committed:
bluetooth: gatt: add authorization callback API for gatt operations

Added the GATT authorization callback API that allows the user to
define application-specific authorization logic for GATT operations.

Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

Thalley commented:
Is this necessary? 

So far we've just done the authorization in the read/write callbacks, which seems as simple (or simpler as you don't have to know what the `attr` pointer points to). 

What is the reason you went for a new callback structure, rather than the existing `gatt_cb`? Or adding it as part of the individual characteristics, like the current read and write callbacks?

kapi-no commented:
> Is this necessary?
> 
> So far we've just done the authorization in the read/write callbacks, which seems as simple (or simpler as you don't have to know what the `attr` pointer points to).
> 
> What is the reason you went for a new callback structure, rather than the existing `gatt_cb`? Or adding it as part of the individual characteristics, like the current read and write callbacks?

I suggested this new API as it provides a single entry point and a more convenient way (in my opinion) of managing authorization for GATT operations that interact with sensitive data. These data types often require application-specific authorization logic.

I am aware that you can return the authorization ATT error for each characteristic read or write callback (`bt_gatt_attr_read_func_t` and `bt_gatt_attr_read_func_t`). However, you would need to adapt all of the service modules in your application by adding proper callbacks (e.g. DIS service module). Then, the application needs to set up relevant callbacks for every characteristic that requires application-specific authorization. This adaptation also includes characteristics of core services (GAP and GATT service) that are defined in the Bluetooth stack (gatt.c). These core services also require additional APIs to enable application-specific authorization. From the perspective of the API user, this approach seems cumbersome.


I haven't added the authorization callbacks to the `struct bt_gatt_cb` for various reasons.

The `bt_gatt_cb_register` API (used to register `struct bt_gatt_cb` structure) allows the user to register multiple callbacks that are used to notify them about the Bluetooth stack events (like MTU change). However, the `struct bt_gatt_cb` callback API does not require any user input to change Bluetooth stack behaviour. In contrast, the `struct bt_gatt_authorization_cb` does require input from the user (consent to authorize operation as a boolean return value). As a result, the `bt_gatt_authorization_cb_register` allows only one callback to be registered. See the `bt_conn_auth_cb_register` and `bt_conn_auth_info_cb_register` for similar differentiation (conn.h header).

The second reason for separation is memory optimization. Thanks to the `CONFIG_BT_GATT_AUTHORIZATION_CUSTOM` Kconfig and dedicated authorization API, memory usage is minimal for users who do not plan to enable this feature. Most of the feature code (90%) is removed when the `CONFIG_BT_GATT_AUTHORIZATION_CUSTOM` is disabled.

Thalley commented:
Thanks for the nice response (some of this probably should belong in the commit message). 

> I am aware that you can return the authorization ATT error for each characteristic read or write callback (`bt_gatt_attr_read_func_t` and `bt_gatt_attr_read_func_t`). However, you would need to adapt all of the service modules in your application by adding proper callbacks (e.g. DIS service module). Then, the application needs to set up relevant callbacks for every characteristic that requires application-specific authorization. This adaptation also includes characteristics of core services (GAP and GATT service) that are defined in the Bluetooth stack (gatt.c). These core services also require additional APIs to enable application-specific authorization. From the perspective of the API user, this approach seems cumbersome.

That makes sense. This is the only way to really add authorization of those (without adding new callbacks for those specifically). However, the `attr` provided by this callback will similarly be unknown to the application (unless additional effort has been made to find the attr of e.g. Device Name). This applies to all stack-defined characteristics. So given that you'd need to figure out what the `attr` is, I'm still not fully convinced by this approach, but I also do not see an obvious better solution. 

> I haven't added the authorization callbacks to the `struct bt_gatt_cb` for various reasons.
> 
> The `bt_gatt_cb_register` API (used to register `struct bt_gatt_cb` structure) allows the user to register multiple callbacks that are used to notify them about the Bluetooth stack events (like MTU change). However, the `struct bt_gatt_cb` callback API does not require any user input to change Bluetooth stack behaviour. In contrast, the `struct bt_gatt_authorization_cb` does require input from the user (consent to authorize operation as a boolean return value). As a result, the `bt_gatt_authorization_cb_register` allows only one callback to be registered. See the `bt_conn_auth_cb_register` and `bt_conn_auth_info_cb_register` for similar differentiation (conn.h header).

That is also a fair point, although you could argue that this is similar to the `bt_conn_cb.le_param_req` callback, i.e. all who have registered this callback will have to approve. The same approach could have been done here. 

> The second reason for separation is memory optimization. Thanks to the `CONFIG_BT_GATT_AUTHORIZATION_CUSTOM` Kconfig and dedicated authorization API, memory usage is minimal for users who do not plan to enable this feature. Most of the feature code (90%) is removed when the `CONFIG_BT_GATT_AUTHORIZATION_CUSTOM` is disabled.

I'm not sure how this relates to the callback structure; you can guard it in a similar way regardless of which way it's added. 

I'm not currently at a point where I'd be ready to either approval or request additional changes yet - I'm neither for or against the suggested changes as of right now. I'll let other people chime in as well before making up my mind :)

jori-nordic commented on include/zephyr/bluetooth/gatt.h:
```diff
@@ -377,6 +408,17 @@ struct bt_gatt_cpf {
+ *
+ *  @return Zero on success or negative error code otherwise
+ */
+int bt_gatt_authorization_cb_register(const struct bt_gatt_authorization_cb *cb);
```
can you document that you first have to clear the cb by passing in NULL? E.g. if you want to change the current cb to point to something else.

Also this is missing something about why this is necessary.
I would change L413-L414 to
```
 *  Register callbacks to perform application-specific authorization of
 *  GATT operations on all characteristics and UUIDs.
```

kapi-no replied:
Sure, I will improve the API documentation with the next patch. I also forgot to mention here that you need to enable the `CONFIG_BT_GATT_AUTHORIZATION_CUSTOM` Kconfig to use this API

kapi-no committed:
tests: bsim: bluetooth: host: gatt: add authorization callback API test

Added a new BabbleSim test that validates the authorization callback API
from the Bluetooth GATT header.

Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

kapi-no commented:
In the last patch, I added a BabbleSim test for the new API and improved the new API documentation in the GATT header.

kapi-no commented on tests/bsim/bluetooth/host/gatt/authorization/src/gatt_server_test.c:
```diff
@@ -0,0 +1,370 @@
+			       NULL, write_cp_chrc, NULL),
+);
+
+static bool gatt_read_operation_authorize(struct bt_conn *conn,
+					  const struct bt_gatt_attr *attr)
+{
+	if (bt_uuid_cmp(attr->uuid, TEST_UNAUTHORIZED_CHRC_UUID) == 0) {
+		unauthorized_chrc_ctx.auth_read_cnt++;
+		return false;
+	} else if (bt_uuid_cmp(attr->uuid, TEST_AUTHORIZED_CHRC_UUID) == 0) {
+		authorized_chrc_ctx.auth_read_cnt++;
+		return true;
+	} else {
+		return true;
+	}
+}
+
+static bool gatt_write_operation_authorize(struct bt_conn *conn,
+					   const struct bt_gatt_attr *attr)
+{
+	if (bt_uuid_cmp(attr->uuid, TEST_UNAUTHORIZED_CHRC_UUID) == 0) {
+		unauthorized_chrc_ctx.auth_write_cnt++;
+		return false;
+	} else if (bt_uuid_cmp(attr->uuid, TEST_AUTHORIZED_CHRC_UUID) == 0) {
+		authorized_chrc_ctx.auth_write_cnt++;
+		return true;
+	} else {
+		return true;
+	}
+}
```
Yes, it also works for the GAP device name (see the example usage in the PR description) and all GATT attributes that are registered

kapi-no commented on include/zephyr/bluetooth/gatt.h:
```diff
@@ -235,6 +235,37 @@ struct bt_gatt_cb {
 };
 
+/** @brief GATT authorization callback structure. */
+struct bt_gatt_authorization_cb {
+	/** @brief Authorize the GATT read operation.
+	 *
+	 *  This callback allows the application to authorize the GATT
+	 *  read operation for the attribute that is being read.
+	 *
+	 *  @param conn Connection object.
+	 *  @param attr The attribute that is being read.
+	 *
+	 *  @retval true  Authorize the operation and allow it to execute.
+	 *  @retval false Reject the operation and prevent it from executing.
+	 */
+	bool (*read_operation_authorize)(struct bt_conn *conn,
+					 const struct bt_gatt_attr *attr);
+
+	/** @brief Authorize the GATT write operation.
+	 *
+	 *  This callback allows the application to authorize the GATT
+	 *  write operation for the attribute that is being written.
+	 *
+	 *  @param conn Connection object.
+	 *  @param attr The attribute that is being written.
+	 *
+	 *  @retval true  Authorize the operation and allow it to execute.
+	 *  @retval false Reject the operation and prevent it from executing.
+	 */
+	bool (*write_operation_authorize)(struct bt_conn *conn,
+					  const struct bt_gatt_attr *attr);
+};
```
It is a security callback that is not purely informational but requires input from the application. It is similar to the `bt_conn_auth_cb_register` and is intended to have one owner. The idea is to have one central module at the application layer that manages the authorization of various operations.

However, it is still possible to implement the dispatching module at the application level if the user wants to define multiple callbacks. I wouldn't overload the Bluetooth stack with additional logic that is related to supporting multiple callbacks. It is very hard to predict the user requirements and the way of handling disputes when two different callback sets return contradictory values for the same GATT attribute.

The registration of multiple callbacks in the `le_param_req` is poorly adopted by applications. See the description in the API documentation that discourages the user from having multiple callbacks: "It is recommended for an application to have just one of these callbacks for simplicity."

kapi-no commented on subsys/bluetooth/host/att.c:
```diff
@@ -131,6 +131,11 @@ static uint16_t bt_att_mtu(struct bt_att_chan *chan)
+/* Descriptor of application-specific authorization callbacks that are used
+ * with the CONFIG_BT_GATT_AUTHORIZATION_CUSTOM Kconfig enabled.
+ */
+const static struct bt_gatt_authorization_cb *authorization_cb;
```
Yes, it is removed during linking. I try to avoid ifdefs when it is possible as they make code harder to read

Thalley replied:
Interesting. I was not aware that `static` variables would be removed during linking, which is why we have so many guarded variables in the first place. 

I'll run a few tests in the LE Audio files to see how many guards and Kconfigs we can actually remove if this is true.

jhedberg replied:
> Interesting. I was not aware that `static` variables would be removed during linking, which is why we have so many guarded variables in the first place.
> 
> I'll run a few tests in the LE Audio files to see how many guards and Kconfigs we can actually remove if this is true.

It's not just that you can remove many guards around static symbols, but once static functions and variables don't have guards around them it opens more possibilities for using `IS_ENABLED()` in the code that accesses these static symbols, i.e. you can then potentially remove #ifdef guards in even more places.

kapi-no commented on subsys/bluetooth/host/att.c:
```diff
@@ -3938,3 +4001,23 @@ bool bt_att_chan_opt_valid(struct bt_conn *conn, enum bt_att_chan_opt chan_opt)
+int bt_gatt_authorization_cb_register(const struct bt_gatt_authorization_cb *cb)
+{
+	if (!IS_ENABLED(CONFIG_BT_GATT_AUTHORIZATION_CUSTOM)) {
+		return -ENOTSUP;
```
Thanks. It seems that ENOSYS fits better here

Thalley reviewed (dismissed):
All my concerns have been answered and I don't see any issues with the implementation. LGTM

kapi-no commented:
Requesting @jori-nordic review again as his approval was dismissed as a result of the recent patch

alwa-nordic commented on tests/bsim/bluetooth/host/gatt/authorization/src/gatt_client_test.c:
```diff
@@ -0,0 +1,340 @@
+
+	printk("Connected to %s\n", addr);
+
+	g_conn = conn;
```
I think you are missing a ref here.

carlescufi replied:
@alwa-nordic should this stop the PR from being merged?

kapi-no replied:
The reference call seems unnecessary in this case. Please see the documentation of the `bt_conn_le_create` that creates a reference that the API user needs to release using the `bt_conn_unref` (as it is done in the disconnected callback).

```
/** @brief Initiate an LE connection to a remote device.
 *
 *  Allows initiate new LE link to remote peer using its address.
 *
 *  The caller gets a new reference to the connection object which must be
 *  released with bt_conn_unref() once done using the object.
```

FYI, the connection logic in this test project was copied from the following project: `tests/bsim/bluetooth/host/gatt/general`

alwa-nordic replied:
@carlescufi Yes. I think this should be resolved first.

@kapi-no I see. So you already have a pointer value with reference ownership in `g_conn`. It's semantically weird that you overwrite it like that. I can elaborate on this if you want. Can you replace this with `__ASSERT_NO_MSG(g_conn == conn)`?

kapi-no replied:
Sure, I can also align it in the source project in a separate PR

kapi-no replied:
addressed

kapi-no replied:
I also opened a fix for the source project in the following PR:

https://github.com/zephyrproject-rtos/zephyr/pull/67837

alwa-nordic reviewed (changes requested):
See comment about references in `g_conn`.

jhedberg commented on include/zephyr/bluetooth/gatt.h:
```diff
@@ -235,6 +235,37 @@ struct bt_gatt_cb {
+	 *  @retval true  Authorize the operation and allow it to execute.
+	 *  @retval false Reject the operation and prevent it from executing.
+	 */
+	bool (*read_operation_authorize)(struct bt_conn *conn,
+					 const struct bt_gatt_attr *attr);
+
+	/** @brief Authorize the GATT write operation.
+	 *
+	 *  This callback allows the application to authorize the GATT
+	 *  write operation for the attribute that is being written.
+	 *
+	 *  @param conn Connection object.
+	 *  @param attr The attribute that is being written.
+	 *
+	 *  @retval true  Authorize the operation and allow it to execute.
+	 *  @retval false Reject the operation and prevent it from executing.
+	 */
+	bool (*write_operation_authorize)(struct bt_conn *conn,
+					  const struct bt_gatt_attr *attr);
```
Does "operation" add anything additional information here? I'd have just called them `read_authorize` (or `authorize_read`), etc.

jhedberg replied:
> Does "operation" add any additional information here? I'd have just called them `read_authorize` (or `authorize_read`), etc.

@kapi-no @jori-nordic any comment on the above?

Thalley replied:
^ I could be in favor of that small rename

jori-nordic replied:
yeah `operation` is superfluous, you can remove it imo

kapi-no replied:
yes, we can remove the `_operation_` infix
